src/daft-micropartition/src/micropartition.rs

```diff
@@ -340,6 +347,47 @@ fn materialize_scan_task(
                         Ok(vec![table])
                     })?
                 }
+                FileFormatConfig::PythonFunction => {
+                    let tables = scan_task.sources.iter().map(|source| {
+                        // Materialize the Tables by calling the Python function
+                        match source {
+                            DataFileSource::PythonFactoryFunction {
+                                func_import_path,
+                                func_args,
+                                ..
+                            } => {
+                                Python::with_gil(|py| {
+                                    let split_path = func_import_path.split('.').collect::<Vec<_>>();
+                                    let (module_path, func_name) = match split_path.as_slice() {
+                                        [module @ .., last] => (module.join("."), last),
+                                        _ => panic!("Cannot import from invalid function at path: {func_import_path}")
+                                    };
+
+                                    let func = py.import(pyo3::types::PyString::new(py, &module_path)).unwrap_or_else(|_| panic!("Cannot import factory function from module {module_path}")).getattr(pyo3::types::PyString::new(py, func_name)).unwrap_or_else(|_| panic!("Cannot find function {func_name} in module {module_path}"));
+                                    let pytables = func.call(func_args.to_pytuple(py), None)?.extract::<Vec<daft_table::python::PyTable>>();
+                                    pytables.map(|pytable_vec| pytable_vec.into_iter().map(|t| t.table))
+                                })
+                            },
+                            _ => unreachable!("PythonFunction file format must be paired with PythonFactoryFunction data file sources"),
+                        }
+                    }).flat_map(|result| match result {
+                        // Apply filters and limits to each Table
+                        Ok(pytables) => pytables.map(|mut tbl| {
+                            if let Some(filters) = scan_task.pushdowns.filters.as_ref() {
+                                tbl = tbl.filter(&[filters.clone()])?;
+                            }
+                            if let Some(limit) = scan_task.pushdowns.limit {
```

Comment:
Instead of a limit on each table, you should do a `take_while` until the limit has been reached.

Reply:
This, paired with the function returning an iterator will let us skip work.

Reply:
Couldn't abstract it all behind an iterator because of some lifetime wrangling required with `'py`, but I used loops instead which should be the same thing but hopefully a little easier to read.